### ISSUE TITLE
how to add new layers to a model

### ISSUE THREAD

**Issue by benwu232** [fchollet/keras#3259]
Hi,

For example, I'd like to insert some new layers to VGG model before the dense layers, load the parameters, freeze them and continue training.

I followed some old issues, which are popping up the top dense and outupt layers, adding new layers and the dense and output layers again. 

When I try to add new layers, I failed with this error:
Exception: Input 0 is incompatible with layer pool6_zp: expected ndim=4, found ndim=2

Actually I also tried just add back the flatten layer and failed too.
I'm using Keras 1.0.4 with tensorflow as backend.

I'm really confused. How can I do that?

Thanks a lot!

Here is the test code:

```python
from keras.layers import (AveragePooling2D, Convolution2D, Dense, Dropout,
    Flatten, Input, MaxPooling2D, ZeroPadding2D)
from keras.models import Model, Sequential

vgg_model = Sequential()
vgg_model.add(Convolution2D(64, 3, 3, activation='relu', input_shape=(3, 224, 224), border_mode='same', name='conv1_1'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv1_1_zp'))
vgg_model.add(Convolution2D(64, 3, 3, activation='relu', name='conv1_2'))
vgg_model.add(MaxPooling2D((2, 2), strides=(2, 2), name='pool1'))

vgg_model.add(ZeroPadding2D((1, 1), name='pool1_zp'))
vgg_model.add(Convolution2D(128, 3, 3, activation='relu', name='conv2_1'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv2_1_zp'))
vgg_model.add(Convolution2D(128, 3, 3, activation='relu', name='conv2_2'))
vgg_model.add(MaxPooling2D((2, 2), strides=(2, 2), name='pool2'))

vgg_model.add(ZeroPadding2D((1, 1), name='pool2_zp'))
vgg_model.add(Convolution2D(256, 3, 3, activation='relu', name='conv3_1'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv3_1_zp'))
vgg_model.add(Convolution2D(256, 3, 3, activation='relu', name='conv3_2'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv3_2_zp'))
vgg_model.add(Convolution2D(256, 3, 3, activation='relu', name='conv3_3'))
vgg_model.add(MaxPooling2D((2, 2), strides=(2, 2), name='pool3'))

vgg_model.add(ZeroPadding2D((1, 1), name='conv4_1_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv4_2'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv4_2_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv4_3'))
vgg_model.add(MaxPooling2D((2, 2), strides=(2, 2), name='pool4'))

vgg_model.add(ZeroPadding2D((1, 1), name='pool4_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv5_1'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv5_1_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv5_2'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv5_2_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv5_3'))
vgg_model.add(MaxPooling2D((2, 2), strides=(2, 2), name='pool5'))

vgg_model.add(Flatten())
vgg_model.add(Dense(4096, activation='relu', name='fc6'))
vgg_model.add(Dropout(0.5, name='fc6do'))
vgg_model.add(Dense(4096, activation='relu', name='fc7'))
vgg_model.add(Dropout(0.5, name='fc7do'))
vgg_model.add(Dense(1000, activation='softmax', name='vgg_classes'))

del vgg_model.layers[-6:]
'''
vgg_model.add(ZeroPadding2D((1, 1), name='pool6_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv6_1'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv6_1_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv6_2'))
vgg_model.add(ZeroPadding2D((1, 1), name='conv6_2_zp'))
vgg_model.add(Convolution2D(512, 3, 3, activation='relu', name='conv6_3'))
vgg_model.add(MaxPooling2D((2, 2), strides=(2, 2), name='pool5'))
'''
vgg_model.add(Flatten())
vgg_model.add(Dense(4096, activation='relu', name='fc6'))
vgg_model.add(Dropout(0.5, name='fc6do'))
vgg_model.add(Dense(4096, activation='relu', name='fc7'))
vgg_model.add(Dropout(0.5, name='fc7do'))
vgg_model.add(Dense(1000, activation='softmax', name='vgg_classes'))


```

**Comment by benwu232** [comment]
Just found this is duplicated with #2640 

Thanks @joelthchao

**Comment by alyato** [comment]
@benwu232 
 I wanna use this pretrained vgg19.  

>  x = Flatten(name='flatten')(x)
        x = Dense(4096, activation='relu', name='fc1')(x)
        x = Dense(4096, activation='relu', name='fc2')(x)
        x = Dense(1000, activation='softmax', name='predictions')(x)
 model = Model(img_input, x)
    return model

 i have only 8 labels to classlify. 
 But it is used by the keras function API ,not the `Sequential` model.
 how can i use the `model.layers.pop()`
like this

>  x = Dense(1000, activation='softmax', name='predictions')(x)
model = Model(img_input, x)
model.layers.pop()
 x = Dense(8, activation='softmax', name='predictions')(x)
 loading weight
return model

Do you give me some advices? Thanks

**Comment by srv902** [comment]
@alyato 

After `model.layers.pop()`, you can do the following

`x = model.layers[-1].output`
`x = Dense(8, activation='softmax', name='predictions')(x)`
`model1 = Model(input=img_input,output=x)`
`return model1`

-1 in `model.layers[-1].output` is the last layer before the softmax layer with 1000 nodes.

**Comment by Jdugand** [comment]
Hello 
I would like to know if when you add a new layer as explained above, you shall compile the model or is optional?
I'm not fully sure of the Compile API scope.

In other words, if when I have a pre-trained model and I add more layers, then
1. what happens if I compile the model and
2. what happens if I don't.

Thanks in advance for your support :-)
JD

**Comment by JanderHungrige** [comment]
@Jdugand 
Late to the party but you need to compile your model if you want to train it. 
If you saved your model compiled and you do not change the model, you do not have to recompile. But as soon as you change something, or just saved the architecture and weights you have to recompile. 
If you just want to use the model for inference, you do not have to recompile.
